undertow-servlet/src/test/java/org/wildfly/elytron/web/undertow/server/servlet/FormAuthenticationSSOBase.java

```diff
@@ -156,41 +158,67 @@ public void testSingleSignOnAcrossTwoAppsWithLogout() throws Exception {
         assertSuccessfulResponse(execute, "ladybird");
         String appOneSessionId = getSessionIdForApp(cookieStore, this::getContextRootAppA);
+        assertNotNull(appOneSessionId);
 
         // can now access APP_B without logging in again
         assertSuccessfulResponse(httpClient.execute(new HttpGet(createUriAppB(null))), "ladybird");
         String appTwoSessionId = getSessionIdForApp(cookieStore, this::getContextRootAppB);
+        assertNotNull(appTwoSessionId);
 
         // log out of APP_A
         httpClient.execute(new HttpGet(createUriAppA("/logout")));
 
         // log into APP_A again
-        execute = loginToApp(httpClient, this::createUriAppA, "ladybird", "Coleoptera");
+        execute = loginToApp(httpClient, this::createUriAppA, "ladybird", "Coleoptera", false);
         assertTrue(cookieStore.getCookies().stream().filter(cookie -> cookie.getName().equals("JSESSIONSSOID")).findAny().isPresent());
         assertSuccessfulResponse(execute, "ladybird");
         String appOneNewSessionId = getSessionIdForApp(cookieStore, this::getContextRootAppA);
 
         // the session ID for APP_A should now be different from the initial session ID
-        assertTrue(appOneSessionId != null && appOneNewSessionId != null && ! appOneSessionId.equals(appOneNewSessionId));
+        assertNotNull(appOneNewSessionId);
+        assertTrue(! appOneSessionId.equals(appOneNewSessionId));
 
         // access APP_B without logging in again
         assertSuccessfulResponse(httpClient.execute(new HttpGet(createUriAppB(null))), "ladybird");
         String appTwoNewSessionId = getSessionIdForApp(cookieStore, this::getContextRootAppB);
 
         // the session ID for APP_B should now be different from the initial session ID
-        assertTrue(appTwoSessionId != null && appTwoNewSessionId != null && ! appTwoSessionId.equals(appTwoNewSessionId));
+        assertNotNull(appTwoNewSessionId);
+        assertTrue(! appTwoSessionId.equals(appTwoNewSessionId));
+
+        // Now try re-authentication
+
+        // log into APP_B as a new user
+        execute = loginToApp(httpClient, this::createUriAppB, "dung", "Coleopterida", true);
+        assertTrue(cookieStore.getCookies().stream().filter(cookie -> cookie.getName().equals("JSESSIONSSOID")).findAny().isPresent());
+        assertSuccessfulResponse(execute, "dung");
+        String appTwoNewNewSessionId = getSessionIdForApp(cookieStore, this::getContextRootAppB);
+
+        // Verify that the session ID changed again.
+        assertNotNull(appTwoNewNewSessionId);
+        assertTrue(! appTwoNewNewSessionId.equals(appTwoNewSessionId));
+
+        // Access App A without logging in again
+        assertSuccessfulResponse(httpClient.execute(new HttpGet(createUriAppA(null))), "dung");
+        String appOneNewNewSessionId = getSessionIdForApp(cookieStore, this::getContextRootAppB);
```

Comment:
Should this be `getContextRootAppA` instead of `getContextRootAppB`?

Reply:
Good catch, now the test fails :smiley: looking more closely now.